pkg/util/purge/purge.go

```diff
@@ -0,0 +1,323 @@
+//     - the role assignment belongs to one of RG: v4-westeurope, v4-eastus, v4-australiasoutheast
+func (rc *ResourceCleaner) cleanRoleAssignments(ctx context.Context) error {
+	roleAssignments, err := rc.roleassignmentcli.List(ctx, "")
+	if err != nil {
+		return err
+	}
+
+	// iterate over all roleAssignments and delete everything not associated with keep cluster
+	for roleAssignments.NotDone() {
+		pageResults := roleAssignments.Values()
+
+		for _, ra := range pageResults {
+			_, err := rc.resourcegroupscli.Get(ctx, *ra.Name)
+			if err != nil {
+				if detailedErr, ok := err.(autorest.DetailedError); ok &&
+					detailedErr.StatusCode == http.StatusNotFound {
+					continue
+				}
+				return err
+			}
+
+			_, err = rc.serviceprincipalcli.Get(ctx, *ra.PrincipalID)
+			if err != nil {
+				if detailedErr, ok := err.(autorest.DetailedError); ok &&
+					detailedErr.StatusCode == http.StatusNotFound {
+					continue
+				}
+				return err
+			}
+
+			// skip if the role assignment is not in the defined resource group
+			match := false
+			for _, region := range rc.resourceRegions {
+				if strings.HasSuffix(*ra.Scope, fmt.Sprintf("%s/providers/Microsoft.Network/virtualNetworks/dev-vnet", region)) {
+					match = true
+				}
+			}
+			if !match {
+				continue
+			}
+
+			if !rc.dryRun {
+				_, err := rc.roleassignmentcli.DeleteByID(ctx, *ra.ID)
+				if err != nil {
+					return err
+				}
+			}
+			rc.log.Infof("Deleting role assignment: %s", *ra.Name)
+		}
+
+		err := roleAssignments.NextWithContext(ctx)
+		if err != nil {
+			return err
+		}
+	}
+	return nil
+}
+
+// CleanApps lists all apps from the subscription and clear unused
+// the deletion works by:
+//     - checking whether coresponding resource group exists (the name of app and resource group matches)
+//     - if there is no existing resource group and the application url has prefix: 'https://az.aro.azure.com/' delete
+//     - delete coresponding service principal
+func (rc *ResourceCleaner) cleanApps(ctx context.Context) error {
+	apps, err := rc.applicationscli.List(ctx, "")
+	if err != nil {
+		rc.log.Errorf("Cannot list apps from subscription: %s", err)
+		return err
+	}
+
+	for apps.NotDone() {
+		pageResult := apps.Values()
+
+		for _, app := range pageResult {
+			_, err := rc.resourcegroupscli.Get(ctx, *app.DisplayName)
+			if detailedErr, ok := err.(autorest.DetailedError); ok &&
+				detailedErr.StatusCode == http.StatusNotFound {
+				continue
+			}
+
+			if len(*app.IdentifierUris) != 1 || !strings.HasPrefix((*app.IdentifierUris)[0], "https://az.aro.azure.com") {
+				continue
+			}
+
+			// get service principal attached to this application
+			sp, err := rc.applicationscli.GetServicePrincipalsIDByAppID(ctx, *app.AppID)
+			if err != nil {
+				return err
+			}
+			rc.log.Infof("Deleting service principal ID: %s", *sp.Value)
+
+			if !rc.dryRun {
+				_, err = rc.serviceprincipalcli.Delete(ctx, *sp.Value)
+				if err != nil {
+					return err
+				}
```

Comment:
I think deleting an AAD app cascades deletions to its SPs, so this should not be needed.